feat(new): Added Azure.VNET.PrivateSubnet (#2999)
Azure/PSRule.Rules.Azure: docs/en/rules/Azure.VNET.PrivateSubnet.md

- Security: Following the zero trust network security principle, it is not recommended to expose a virtual network to the internet by default.
- Explicit Connectivity: It is better to use explicit methods of connectivity rather than implicit ones for granting internet access to VMs.
- Ownership and Stability: The default outbound access IP is managed by Microsoft, and its ownership or address might change, potentially causing disruptions.

Enabling the private subnet feature on a subnet prevents VMs within that subnet from using default outbound access to connect to public endpoints.

This also applies to VMs within a scale set configured with uniform orchestration mode.

Recommended approach:

- Use explicit methods for public connectivity, such as:
  - NAT Gateway: Provides a stable and owned public IP for outbound connectivity.
  - Firewall: Offers outbound connectivity with built-in NAT capabilities.

Note that there are limitations to this feature, so please refer to the documentation for detailed information.

## RECOMMENDATION

Consider using an explicit method of public connectivity for virtual machines.

## EXAMPLES

### Configure with Azure template

> For each subnet object in defined the `properties.subnets` property:
>   - Set the `properties.defaultOutboundAccess` property to `false`.

Reviewer: ```suggestion
To configure virtual networks that pass this rule:

- For each subnet in defined the `properties.subnets` property:
  - Set the `properties.defaultOutboundAccess` property to `false`.
```

For example:

```json
{
  "type": "Microsoft.Network/virtualNetworks",
  "apiVersion": "2023-11-01",
  "name": "[parameters('name')]",
  "location": "[parameters('location')]",
  "properties": {
    "addressSpace": {
      "addressPrefixes": [
        "10.0.0.0/16"
      ]
    },
    "subnets": [
      {
        "name": "subnet-A",
        "properties": {
          "addressPrefix": "10.0.0.0/27",

>            "defaultOutboundAccess": false

Reviewer: ```suggestion
          "defaultOutboundAccess": false
```

        }
      },
      {
        "name": "subnet-B",
        "properties": {
          "addressPrefix": "10.0.0.32/27",
          "defaultOutboundAccess": false
        }
      }
    ]
  }
}
```

### Configure with Bicep

> To configure virtual networks that pass this rule:
> 
> For each subnet object in defined the `properties.subnets` property:
>   - Set the `properties.defaultOutboundAccess` property to `false`.

Reviewer: ```suggestion
To configure virtual networks that pass this rule:

- For each subnet in defined the `properties.subnets` property:
  - Set the `properties.defaultOutboundAccess` property to `false`.
```

For example:

```bicep
resource vnet 'Microsoft.Network/virtualNetworks@2023-11-01' = {
  name: name
  location: location
  properties: {
    addressSpace: {
      addressPrefixes: [
        '10.0.0.0/16'
      ]
    }
    subnets: [
      {
        name: 'subnet-A'
        properties: {
          addressPrefix: '10.0.0.0/27'
          defaultOutboundAccess: false
        }
      }
      {
        name: 'subnet-B'
        properties: {
          addressPrefix: '10.0.0.32/27'
          defaultOutboundAccess: false
        }
      }
    ]
  }
}
```

## NOTES

This feature is currently in preview.

## LINKS

> - [SE:06 Network controls](https://learn.microsoft.com/azure/well-architected/security/networking)
> - [Default outbound access](https://learn.microsoft.com/azure/virtual-network/ip-services/default-outbound-access)
> - [Azure deployment reference - Virtual Network](https://learn.microsoft.com/azure/templates/microsoft.network/virtualnetworks)
> - [Azure deployment reference - Subnet](https://learn.microsoft.com/azure/templates/microsoft.network/virtualnetworks/subnets)

Reviewer: ```suggestion
- [SE:06 Network controls](https://learn.microsoft.com/azure/well-architected/security/networking)
- [Default outbound access](https://learn.microsoft.com/azure/virtual-network/ip-services/default-outbound-access)
- [Plan for inbound and outbound internet connectivity](https://learn.microsoft.com/azure/cloud-adoption-framework/ready/azure-best-practices/plan-for-inbound-and-outbound-internet-connectivity)
- [Azure deployment reference - Virtual Network](https://learn.microsoft.com/azure/templates/microsoft.network/virtualnetworks)
- [Azure deployment reference - Subnet](https://learn.microsoft.com/azure/templates/microsoft.network/virtualnetworks/subnets)
```
